ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleTable.scala

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.backend.oracle
+
+import java.sql.Connection
+import com.daml.platform.store.backend.common.{BaseTable, Field, Table}
+
+private[oracle] object OracleTable {
+  private def idempotentBatchedInsertBase[FROM](
+      insertStatement: String,
+      keyFieldIndex: Int,
+  )(fields: Seq[(String, Field[FROM, _, _])]): Table[FROM] =
+    new BaseTable[FROM](fields) {
+      override def executeUpdate: Array[Array[_]] => Connection => Unit =
+        data =>
+          connection =>
+            Table.ifNonEmpty(data) {
+              val preparedStatement = connection.prepareStatement(insertStatement)
+              data(0).indices.foreach { dataIndex =>
+                fields(keyFieldIndex)._2.prepareData(
+                  preparedStatement,
+                  1,
+                  data(keyFieldIndex)(dataIndex),
+                )
+                fields.indices.foreach { fieldIndex =>
+                  preparedStatement.setObject(fieldIndex + 2, data(fieldIndex)(dataIndex))
+                }
+                preparedStatement.addBatch()
+              }
+              preparedStatement.executeBatch()
+              preparedStatement.close()
+            }
+    }
+
+  private def idempotentBatchedInsertStatement(
+      tableName: String,
+      fields: Seq[(String, Field[_, _, _])],
+      keyFieldIndex: Int,
+  ): String = {
+    def commaSeparatedOf(extractor: ((String, Field[_, _, _])) => String): String =
+      fields.view
+        .map(extractor)
+        .mkString(",")
+    val tableFields = commaSeparatedOf(_._1)
+    val selectFields = commaSeparatedOf { case (_, field) =>
+      field.selectFieldExpression("?")
+    }
+    val keyFieldName = fields(keyFieldIndex)._1
+    val keyFieldSelectExpression = fields(keyFieldIndex)._2.selectFieldExpression("?")
+    s"""MERGE INTO $tableName USING DUAL on ($keyFieldName = $keyFieldSelectExpression)
```

Comment:
is this the same as in H2? if yes we could just put this to common and make the default by viring to the common FieldStrategy

Reply:
oracle needs parentheses around hte `($keyFieldName = $keyFieldSelectExpression)`, h2 does not : otherwise they should be equiv here- i'll try to extract out a common version that works for both

Reply:
I'd rather leave this to refactoring as a follow-on PR as it involves some moving around of Table and Schema common code